tests/test_auto_tokenize.py

```diff
@@ -0,0 +1,209 @@
+import pytest
+
+from mpi4py import MPI
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+
+comm = MPI.COMM_WORLD
+rank = comm.Get_rank()
+size = comm.Get_size()
+
+
+def test_allreduce():
+    from mpi4jax import allreduce, auto_tokenize
+
+    arr = jnp.ones((3, 2))
+
+    def f(a):
+        res, _ = allreduce(a, op=MPI.SUM)
+        res, _ = allreduce(res, op=MPI.SUM)
+        return res
+
+    res = auto_tokenize(f)(arr)
+    np.testing.assert_allclose(res, arr * size ** 2)
+
+
+def test_nested_jits():
+    from mpi4jax import allreduce, auto_tokenize
+
+    arr = jnp.ones((3, 2))
+
+    @jax.jit
+    def f(a):
+        res, _ = allreduce(a, op=MPI.SUM)
+
+        @jax.jit
+        def g(b):
+            res, _ = allreduce(b, op=MPI.SUM)
+            res, _ = allreduce(res, op=MPI.SUM)
+            return res
+
+        res = g(res)
+        res = g(res)
+        return res
+
+    res = auto_tokenize(f)(arr)
+    np.testing.assert_allclose(res, arr * size ** 5)
+
+
+def test_nested_jits():
+    from mpi4jax import allreduce, auto_tokenize
+
+    arr = jnp.ones((3, 2))
+
+    @jax.jit
+    def f(a):
+        res, _ = allreduce(a, op=MPI.SUM)
+
+        @jax.jit
+        def g(b):
+            res, _ = allreduce(b, op=MPI.SUM)
+            res, _ = allreduce(res, op=MPI.SUM)
+            return res
+
+        res = g(res)
+        res = g(res)
+        return res
+
+    res = auto_tokenize(f)(arr)
+    np.testing.assert_allclose(res, arr * size ** 5)
+
+
+@pytest.mark.skipif(size < 2, reason="need 2 processes")
+def test_send_recv_tokenizer():
+    from mpi4jax import recv, send, auto_tokenize
+
+    def simple_message_pass(arr):
+        # Here, we test a crazy send/recv pattern that is extremely likely to deadlock unless
```

Comment:
You and I have different definitions of crazy :)

Reply:
🔥🥔

Reply:
Actually that was just my obscure way to point out that you use the "crazy" comment for the simple test case, too.

Reply:
oh right...